src/xref_runner.erl

```diff
@@ -0,0 +1,158 @@
+%% -------------------------------------------------------------------
+%% This module is basically copied from rebar's rebar_xref, which in turn
+%% borrows heavily from rebar's  http://github.com/etnt/eCheck project as
+%% written by Torbjorn Tornkvist <tobbe@kreditor.se>, Daniel Luna
+%% <daniel@lunas.se> and others.
+%% -------------------------------------------------------------------
+-module(xref_runner).
+-author('elbrujohalcon@inaka.net').
+
+-type check() :: undefined_function_calls
+               | undefined_functions
+               | locals_not_used
+               | exports_not_used
+               | deprecated_function_calls
+               | deprecated_functions.
+
+-type xref_default() :: {builtins | recurse | verbose | warnings, boolean()}.
+
+-type config() :: #{ extra_paths    => [file:name_all()]
+                   , xref_defaults  => [xref_default()]
+                   , dirs           => [file:name_all()]
+                   }.
+-type warning() :: #{ filename      => file:name_all()
+                    , line          => non_neg_integer()
+                    , source        => mfa()
+                    , target        => mfa()
+                    }.
+
+-export_type([check/0, xref_default/0, config/0, warning/0]).
+-export([check/2]).
+
+-spec check(check(), config()) -> [warning()].
+check(Check, Config) ->
+  XrefDefaults = maps:get(xref_defaults, Config, []),
+  Dirs = maps:get(dirs, Config, ["ebin"]),
+
+  {ok, Xref} = xref:start(?MODULE),
+  try
+    ok = xref:set_library_path(Xref, code_path(Config)),
+
+    xref:set_default(Xref, XrefDefaults),
+
+    lists:foreach(
+      fun(Dir) ->
+        {ok, _} = xref:add_directory(Xref, Dir)
+      end, Dirs),
+
+    {ok, Results} = xref:analyze(Xref, Check),
+
+    FilteredResults = filter_xref_results(Check, Results),
+
+    [result_to_warning(Result) || Result <- FilteredResults]
+  after
+    stopped = xref:stop(Xref)
+  end.
+
+%% ===================================================================
+%% Internal functions
+%% ===================================================================
+
+code_path(Config) ->
+  ExtraPaths = maps:get(extra_paths, Config, []),
+  [P || P <- code:get_path() ++ ExtraPaths, filelib:is_dir(P)].
+
+filter_xref_results(Check, Results) ->
+  SourceModules =
+    lists:usort([source_module(Result) || Result <- Results]),
+
+  Ignores =
+    lists:flatmap(
+      fun(Module) -> get_ignorelist(Module, Check) end, SourceModules),
+
+  [Result || Result <- Results,
+             not lists:member(parse_xref_result(Result), Ignores)].
+
+source_module({Mt,_Ft,_At}) -> Mt;
+source_module({{Ms,_Fs,_As},_Target}) -> Ms.
+
+%%
+%% Ignore behaviour functions, and explicitly marked functions
+%%
+%% Functions can be ignored by using
+%% -ignore_xref([{F, A}, {M, F, A}...]).
+get_ignorelist(Mod, Check) ->
+  %% Get ignore_xref attribute and combine them in one list
+  Attributes =
+    try
+      Mod:module_info(attributes)
```

Comment:
Remove the dynamic function call on line 88. Only modules that define callbacks should make dynamic calls.